### ISSUE TITLE
Leverage ONNF efforts to move as many ONNX operators be generic function calls as possible?

### ISSUE THREAD

**Issue by linkerzhang** [clang-ykt/ONNF#8]
When creating ONNX dialect and defining ONNX operators, I'm seeing an opportunity to go thru ONNX ops and split them into primitive ops and generic function calls, so that the primitive op set will be fairly stable.  @tjingrant what do you think please? Thank you!

**Comment by tjingrant** [comment]
@linkerzhang there are definitely opportunities to exploit this strategy you mentioned. In fact when we were planning for the ONNX dialect, there are operations that clearly need to be defined in terms of other operations (RNN operations for example). Having two categories of operations (composite and elemental operations) make a lot of sense. We can also implement a lowering pass within ONNX dialect to lower composite operations into elemental ones. How do you wish to proceed with this idea? Let me know how I can help.

CC'ing @chentong319 in this discussion.

**Comment by linkerzhang** [comment]
"We can also implement a lowering pass within ONNX dialect to lower composite operations into elemental ones." Exactly! To me, this level of lowering is very meaningful to ONNX.

Let's sync on how this can be done efficiently. I'd understand more about the engineering efforts that you guys will offer so that I'll plan it accordingly with my team. An offline call may be better for the sync, btw.

**Comment by tjingrant** [comment]
It would be tricky to have offline calls these days, as it's the holiday season. Let me send an email to see how to move this forward.

**Comment by tjingrant** [comment]
@linkerzhang how's your next week's schedule looking? A call will be possible with some of us from IBM Research starting next Monday.

**Comment by linkerzhang** [comment]
sounds good. I'm available starting from Tuesday.

**Comment by linkerzhang** [comment]
Circle back. 

@tjingrant I'm seeing that the onnx.md is a 1:1 mapping to ONNX operators.md, instead of having a primitive op set identified + functions, am I misunderstanding sth?

One more question please. are you planning to move this to ONNX org? as we're also working on it, I'm thinking of merging the efforts instead of creating a new repo/efforts under ONNX org by us separately. 

Thank you!

**Comment by tjingrant** [comment]
Yes, ONNF prioritizes being a faithful reflection of ONNX standard. Nonetheless, Implementation-wise, we do tend to break large ops down to smaller primitives (e.g., reduction ops).

Replied to your second question with email, basically I haven't got the bandwidth to look at this issue, but will move forward ASAP.

**Comment by bhack** [comment]
@linkerzhang Please, if you have, give also any feedback for onnxruntime at https://llvm.discourse.group/t/development-of-high-level-tensor-compute-primitives-dialect-s-and-transformations/388/57

**Comment by tungld** [comment]
Regarding "break large ops down to smaller primitives", we are working on this PR #87 which allows rewriting composite operations into element ones in a declarative manner.
